src/stirling/obj_tools/elf_reader_test.cc

```diff
@@ -40,6 +40,54 @@ using ::testing::UnorderedElementsAre;
 
 using ::px::operator<<;
 
+// Models ELF section output information from objdump -h.
+// ELFIO::section's do not contain virtual memory addresses like ELFIO::segment's do
+// so this struct is used to store the information from objdump.
+// Example objdump output:
+//
+// $ objdump -j .bss -h bazel-bin/src/stirling/obj_tools/testdata/cc/test_exe/test_exe
+//
+// bazel-bin/src/stirling/obj_tools/testdata/cc/test_exe/test_exe:     file format elf64-x86-64
+//
+// Sections:
+// Idx Name          Size      VMA               LMA               File off  Algn
+//  27 .bss          00002068  00000000000bd100  00000000000bd100  000ba100  2**5
+//                    ALLOC
+struct Section {
+  std::string name;
+  int64_t size;
+  int64_t vma;
+  int64_t lma;
+  int64_t file_offset;
+};
+
+StatusOr<Section> ObjdumpSectionNameToAddr(const std::string& path,
+                                           const std::string& section_name) {
+  Section section;
+  std::string objdump_out =
+      px::Exec(absl::StrCat("objdump -h -j ", section_name, " ", path)).ValueOrDie();
```

Comment:
Using `px::Exec` breaks the hermeticity of the test. Might be better to run this externally and put the required assets into `testdata`.

Reply:
While there is a hermeticity risk, the expectations are based on the [test_exe target](https://github.com/pixie-io/pixie/blob/50ddcd32eb217e1aa5e87124883ee284a36052a1/src/stirling/obj_tools/testdata/cc/BUILD.bazel#L25), so it's possible that the `test_exe` binary and the objdump output may become out of sync. While the `test_exe` binary could be added as testdata as well, there are other occurrences of accessing the binary directly via the [test_exe_fixture.h header](https://github.com/pixie-io/pixie/blob/50ddcd32eb217e1aa5e87124883ee284a36052a1/src/stirling/obj_tools/testdata/cc/test_exe_fixture.h#L33).
```
$ git grep test_exe_fixture.h
src/stirling/bpf_tools/bcc_wrapper_bpf_test.cc:#include "src/stirling/obj_tools/testdata/cc/test_exe_fixture.h"
src/stirling/obj_tools/elf_reader_test.cc:#include "src/stirling/obj_tools/testdata/cc/test_exe_fixture.h"
src/stirling/obj_tools/testdata/cc/BUILD.bazel:    hdrs = ["test_exe_fixture.h"],
src/stirling/source_connectors/dynamic_bpftrace/dynamic_bpftrace_connector_bpf_test.cc:#include "src/stirling/obj_tools/testdata/cc/test_exe_fixture.h"
src/stirling/source_connectors/dynamic_tracer/dynamic_trace_bpf_test.cc:#include "src/stirling/obj_tools/testdata/cc/test_exe_fixture.h"
```

We also follow this pattern for `nm` [cli output](https://github.com/pixie-io/pixie/blob/50ddcd32eb217e1aa5e87124883ee284a36052a1/src/stirling/obj_tools/elf_reader_test.cc#L46) and haven't had issues with these tests to my knowledge.

My preference would be to keep it as is, but happy to proceed with your suggestion if you feel strongly about it.

Reply:
Just a few thoughts in response:

> "it's possible that the test_exe binary and the objdump output may become out of sync":

If we do check it in, I don't think the risk of them getting out of sync is a concern, because the tests would catch the mismatch.

> "While the test_exe binary could be added as testdata as well, there are other occurrences of accessing the binary directly via the [test_exe_fixture.h header](https://github.com/pixie-io/pixie/blob/50ddcd32eb217e1aa5e87124883ee284a36052a1/src/stirling/obj_tools/testdata/cc/test_exe_fixture.h#L33)."

Yeah, I'm not suggesting to check in `test_exe`. Wouldn't really make sense because it's already in the code-base.

> We also follow this pattern for nm [cli output](https://github.com/pixie-io/pixie/blob/50ddcd32eb217e1aa5e87124883ee284a36052a1/src/stirling/obj_tools/elf_reader_test.cc#L46) and haven't had issues with these tests to my knowledge.

That one's probably not great either :)

Another option is to have bazel generate the testdata by running objdump. That way it stays in sync. But not sure that's worth the hassle.

Having said all that, I don't feel that strongly that you should consider this a blocker. So it's okay by me if you want to put a `TODO` to revisit these two cases in the future, and land as is.

Reply:
I've opened https://github.com/pixie-io/pixie/pull/1980 to update the nm usage to be hermetic. I'm not sure if doing the same for `objdump` will be as easy. (since nm has [built-in custom makefile variable](https://bazel.build/reference/be/make-variables#custom_variables) support, while `objdump` doesn't).

If that ends up being easy to do, I'll fix this otherwise I'll add a `TODO` for later.

Reply:
I've added the TODO and will address this once I hear from James and Vihang regarding how to access `objdump` from the bazel toolchain.